lib/fluent/plugin/filter_kubernetes_metadata.rb

```diff
@@ -274,35 +252,74 @@ def configure(conf)
 
     end
 
-    def get_metadata_for_record(namespace_name, pod_name, container_name)
+    def get_metadata_for_record(match_data, cache_key, create_time)
+      namespace_name = match_data['namespace']
+      pod_name = match_data['pod_name']
       metadata = {
-        'container_name' => container_name,
+        'container_name' => match_data['container_name'],
         'namespace_name' => namespace_name,
-        'pod_name'       => pod_name,
+        'pod_name'       => pod_name
       }
       if @kubernetes_url.present?
-        cache_key = "#{namespace_name}_#{pod_name}"
-
-        this = self
-        pod_metadata = @cache.getset(cache_key) {
-          @stats.bump(:pod_cache_miss)
-          md = this.get_pod_metadata(
-            namespace_name,
-            pod_name,
-          )
-          md
-        }
+        pod_metadata = get_pod_metadata(cache_key, namespace_name, pod_name, create_time)
         metadata.merge!(pod_metadata) if pod_metadata
+      end
+      metadata
+    end
+
+    def get_pod_metadata(key, namespace_name, pod_name, record_create_time)
+        ids = @id_cache.fetch(key) do
+          #no metadata found
+          @stats.bump(:id_cache_miss)
+          pod_metadata = fetch_pod_metadata(namespace_name, pod_name)
+          namespace_metadata = fetch_namespace_metadata(namespace_name)
+          ids = {:pod_id=> pod_metadata['pod_id'], :namespace_id => namespace_metadata['namespace_id'] }
+          # pod not found or namespace not found
+          if ids[:pod_id].nil? || ids[:namespace_id].nil?
```

Comment:
We have four cases of behaviors given the possible results of fetching pod and namespace metadata:

 1. we have pod & namespace metadata
 2. we have pod & not namespace metadata
 3. we have namespace & not pod metadata
 4. we have neither

It is possible due to timing and that the metadata fetch APIs are independent, that case #2 can happen, though not often.  It looks like we are ignoring case 2, but can still return an `ids` variable that has a `pod_id` value with a `nil` `namespace_id` value.  Which is probably bad.

This seems to point at a problem with the Kube APIs, we look up by name and pod, but we don't get back the namespace ID with the pod metadata to allow us to validate which namespace it actually is associated with that namespace (as namespace names can come and go).

Reply:
In this code path we can never have pod metadata without namespace because the client only turns pod meta in context of the namespace.  I would expect a record in this scenario to end up in the orphaned index.

Reply:
@jcantrill, not sure about that; if we have the following timeline

 * t0: fetch_pod_metdata(namespace_name, pod_name)
 * t1: namespace is deleted
 * t2: fetch_namespace_metadata(namespace_name)

Then it seems we have this problem.  If the pod metadata returned the namespace ID, then we'd be able to detect this situation.  Right now, at least from what I understand, we can't.